### PR TITLE
Initial JMX Metrics Extension structure

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
**Description:**
Adding a feature - Adding initial location and component helpers for future JMX Metrics Extension*.  Core functionality is intended to be be added incrementally.

**Link to tracking Issue:** 
~associated JMX Metric Gatherer utility: https://github.com/open-telemetry/community/issues/449~
https://github.com/open-telemetry/opentelemetry-java-contrib/pull/4

**Testing:** base unit tests

**Documentation:** basic readme

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@rmfitzpatrick Can we actually push from the JMX extension that you will have in the java-contrib using the Otel-Java + OTLP exporter directly to the collector?

**Comment by rmfitzpatrick** [comment]
> @rmfitzpatrick Can we actually push from the JMX extension that you will have in the java-contrib using the Otel-Java + OTLP exporter directly to the collector?

Yes, when configured to use the otlp exporter type it stands up an OtlpGrpcMetricExporter to reach the configured endpoint of the otlp receiver: https://github.com/rmfitzpatrick/opentelemetry-java-instrumentation/blob/jmx-metrics/instrumentation-core/jmx-metrics/src/main/java/io/opentelemetry/instrumentation/jmxmetrics/GroovyUtils.java#L56

**Comment by bogdandrutu** [comment]
If just lunches a process I would make this an extension, because it does not have to connect to the processing pipeline, the sub-process will connect via its exporter.

File: receiver/jmxmetricsreceiver/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+partially implemented.
+
+Example configuration:
+
+```yaml
+receivers:
+  jmx_metrics:
+    service_url: service:jmx:rmi:///jndi/rmi://<my-jmx-host>:<my-jmx-port>/jmxrmi
+    groovy_script: /opt/my/groovy.script
+    interval: 10s
```
Comment (jrcamp) [inline receiver/jmxmetricsreceiver/README.md:28]: Usually `collection_interval`

File: receiver/jmxmetricsreceiver/config.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+type config struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"`
+	// The target JMX service url.
+	ServiceURL string `mapstructure:"service_url"`
+	// The script for the metric gatherer to run on the configured interval.
+	GroovyScript string `mapstructure:"groovy_script"`
+	// The endpoint for the metric gatherer to use to report metrics.
+	CollectorEndpoint string `mapstructure:"collector_endpoint"`
```
Comment (jrcamp) [inline receiver/jmxmetricsreceiver/config.go:31]: Do we need to support this? Could be confusing to have metrics flow out of band from a receiver. Can we just always feed the metrics through the pipeline and let the user decide how to send them via exporter?

File: receiver/jmxmetricsreceiver/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+Note: this receiver is in alpha and functionality and configuration fields are subject to change.  It is currently
+partially implemented.
+
+Example configuration:
+
+```yaml
+receivers:
+  jmx_metrics:
+    service_url: service:jmx:rmi:///jndi/rmi://<my-jmx-host>:<my-jmx-port>/jmxrmi
+    groovy_script: /opt/my/groovy.script
```
Comment (jrcamp) [inline receiver/jmxmetricsreceiver/README.md:27]: I think it's more user friendly to be able to put program text directly in yaml if desired.

We need to add a collector feature to be able to reference file paths but once that's solved any component will be able to use it. So I think putting the program text here is better.

**Review by jrcamp — COMMENTED** [review]
In smart agent there was a base Java monitor and a JMX monitor that built on top of that. That way if somebody doesn't want to use Groovy they could still leverage the Java monitor. Should we factor it in a similar way here?

**Comment by bogdandrutu** [comment]
@rmfitzpatrick you agreed that this should be an extension :) Please don't forget to change that

**Comment by rmfitzpatrick** [comment]
> @rmfitzpatrick you agreed that this should be an extension :) Please don't forget to change that
Plan on giving this an earnest go today

**Comment by rmfitzpatrick** [comment]
@bogdandrutu @jrcamp I've moved this to an extension

File: receiver/jmxmetricsreceiver/config.go (hunk 2 above)
Comment (rmfitzpatrick) [reply receiver/jmxmetricsreceiver/config.go]: It's unclear to be how this is fully able to operate w/ the provided utilities in the extension, but I've removed this option to defer the decision until actual functionality is implemented.  otlp and prometheus receivers can be supported, maybe being configured by the receiver creator?

File: extension/jmxmetricsextension/Makefile [hunk 4]
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
Comment (bogdandrutu) [inline extension/jmxmetricsextension/Makefile:1]: Please add the module in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/.github/dependabot.yml

File: extension/jmxmetricsextension/go.mod [hunk 5]
```diff
@@ -0,0 +1,11 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/extension/jmxmetricsextension
+
+go 1.14
+
+require (
+	github.com/stretchr/testify v1.6.1
+	go.opentelemetry.io/collector v0.8.0
```
Comment (bogdandrutu) [inline extension/jmxmetricsextension/go.mod:7]: Please use the same version as everywhere.

**Review by bogdandrutu — APPROVED** [review]

File: extension/jmxmetricsextension/go.mod (hunk 5 above)
Comment (rmfitzpatrick) [reply extension/jmxmetricsextension/go.mod]: updated

**Pull request merged**